messages/src/taskprov.rs

```diff
@@ -0,0 +1,794 @@
+//! Data structures as defined in extension [draft-wang-ppm-dap-taskprov][1].
+//!
+//! [1]: https://datatracker.ietf.org/doc/draft-wang-ppm-dap-taskprov/
+
+use crate::{Duration, Error, Time, Url};
+use anyhow::anyhow;
+use derivative::Derivative;
+use prio::codec::{
+    decode_u16_items, decode_u24_items, decode_u8_items, encode_u16_items, encode_u24_items,
+    encode_u8_items, CodecError, Decode, Encode,
+};
+use std::{
+    fmt::{self, Debug, Formatter},
+    io::Cursor,
+};
+
+/// Defines all parameters necessary to configure an aggregator with a new task.
+/// Provided by taskprov participants in all requests incident to task execution.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct TaskConfig {
+    /// Opaque info specific for a task.
+    task_info: Vec<u8>,
+    /// List of URLs where the aggregator's API endpoints can be found.
+    aggregator_endpoints: Vec<Url>,
```

Comment:
Working through implementing this, not sure if this is the right approach. If we store as a `Vec<url::Url>`, then each of the `encode()`, `encoded_len()`, and `decode()` methods have to convert back to the wire-representation `Url` for each call.

I'm thinking a happy medium might be to `impl TryInto<url::Url> for Url`, that way there's some sugar for directly converting the wire-representation to the richer representation. Use of `Url` in `url::Url` contexts then should just be `url.try_into()?` then, if I'm understanding the `TryInto` trait correctly.

Reply:
`url::Url` implements [`AsRef<str>`](https://docs.rs/url/2.4.0/url/struct.Url.html#impl-AsRef%3Cstr%3E-for-Url), providing "free" conversion of `&url::Url` to `&str`. (There's also [`as_str`](https://docs.rs/url/2.4.0/url/struct.Url.html#method.as_str), which does the same without the trait, and whose documentation & implementation confirm that this is no more expensive than retrieving a reference from the internals of the `url::Url` struct.)

That is, the conversion of `Url` to a byte-string is effectively just grabbing the byte string that the `Url` already stores, i.e. very cheap.

Reply:
Let me know how https://github.com/divviup/janus/pull/1579/commits/6c31966e225f622c3d9739733f09d0f045fe3e11 looks.

Mainly my concern comes around allocating additional Vec's to store the results of conversion while still being able to use `{en,de}code_u16_items()`. It is in the hot-path for Leader usage--each report will need to be decoded. Probably not worth getting too hung up over though.

Reply:
Yeah I think being too hung up on vec allocations would be premature optimization. `url::Url::parse()` will make allocations anyways, so keeping it in the in-memory representation should minimize the amount of `url::Url::parse()` we have to do. The encoding side is unfortunate, but I don't believe that is on the hot path (should only be for encoding the struct for aggregation jobs).

I'll be looking for feedback on semantics/cleanliness in that case.

Reply:
One way to deal with this would be to `impl Encode/Decode on url::Url`, so that we could use `encode_*_items` directly to return a `Vec<url::Url>`. However, this would run afoul of the [orphan rule](https://doc.rust-lang.org/book/ch10-02-traits.html); we could work around this by introducing a newtype for `url::Url`.

But we already have a newtype: we could redefine the message crate's `Url` to be `Url(url::Url)`, implement `Encode` and `Decode` on it, and then use `encode_*_items` to directly get a `Vec<Url>` without allocating an additional vector.

As you say, though, this would be additional complexity and may be premature optimization. Up to you if you want to do this.

Reply:
It turns out to be a pretty nice simplification, at least in terms of code. Great suggestion! https://github.com/divviup/janus/pull/1579/commits/5b03f77be02983c3ca2bdaff0c99e352ccea388b.

The only other caveat is that `url::Url` may make changes to the underlying data, in particular I noticed it will add a trailing slash to `https://example.com`. But I think it's benign in all practical cases--just the encoded length might be longer than what was decoded. I've made a note of this in rustdoc.

Reply:
Nice spot on the automatic transformation -- I think this is OK, too, but it's good to make note of.